OpenSim/Tests/Wrapping/testWrapCylinder.cpp

```diff
@@ -0,0 +1,314 @@
+        WrappingTolerances() = default;
+
+        WrappingTolerances(double eps):
+            position(eps),
+            length(eps)
+        {}
+
+        double position = 1e-9;
+        double length = 1e-9;
+    };
+
+    bool IsEqual(
+        double lhs,
+        double rhs,
+        double tolerance)
+    {
+        return std::abs(lhs - rhs) < tolerance;
+    }
+
+    bool IsEqual(
+        const SimTK::Vec3& lhs,
+        const SimTK::Vec3& rhs,
+        double tolerance)
+    {
+        return IsEqual(lhs[0], rhs[0], tolerance)
+            && IsEqual(lhs[1], rhs[1], tolerance)
+            && IsEqual(lhs[2], rhs[2], tolerance);
+    }
+
+    bool IsEqual(
+        const PathSegment& lhs,
+        const PathSegment& rhs,
+        double tolerance)
+    {
+        return IsEqual(lhs.start, rhs.start, tolerance)
+            && IsEqual(lhs.end, rhs.end, tolerance);
+    }
+
+    bool IsEqual(
+        const WrapTestResult& lhs,
+        const WrapTestResult& rhs,
+        const WrappingTolerances& tolerance)
+    {
+        if (lhs.noWrap && rhs.noWrap) {
+            return true;
+        }
+        return IsEqual(lhs.path, rhs.path, tolerance.position)
+            && IsEqual(lhs.length, rhs.length, tolerance.length)
+            && lhs.noWrap == rhs.noWrap;
+    }
+
+}
+
+// Section on configuring and simulating a specific wrapping scenario.
+namespace {
+
+    // Parameterization of the scenario for testing the wrapping.
+    //
+    // The simulated scene is one with a cylinder at the origin, with a user
+    // defined radius, length, active quadrant, and relative orientation.
+    // A specific wrapping case can be configured by choosing the start and end
+    // points of the path.
+    struct WrapInput final {
+
+        SimTK::Rotation cylinderOrientation() const {
+            SimTK::Rotation orientation;
+            orientation.setRotationToBodyFixedXYZ(eulerRotations);
+            return orientation;
+        }
+
+        // Endpoints of the total path:
+        PathSegment path = {{}, {}};
+
+        // Wrapping cylinder parameters:
+        double radius = 1.;
+        double cylinderLength = 1.;
+        std::string quadrant = "all";
+        SimTK::Vec3 eulerRotations = {0., 0., 0.};
+    };
+
+    std::ostream& operator<<(std::ostream& os, const WrapInput& input) {
+        return os <<
+            "WrapInput{" <<
+            "path: " << input.path << ", " <<
+            "radius: " << input.radius << ", " <<
+            "cylinderLength: " << input.cylinderLength << ", " <<
+            "quadrant: " << input.quadrant << ", " <<
+            "eulerRotations: " << input.eulerRotations << "}";
+    }
+
+    // Simulates the wrapping scenario as configured by the WrapInput.
+    WrapTestResult solve(const WrapInput& input, bool visualize) {
+        Model model;
+        model.setName("testWrapCylinderModel");
+
+        std::unique_ptr<WrapCylinder> cylinder(new WrapCylinder());
```

Comment:
If it doesn't make indentation too ugly, consider a sub-scope that starts before `cylinder` and ends after `cylinder.release()` so that it is logically impossible for code further down to use the released (null) pointers.

Reply:
Aaah, that's what I was looking for!